### DISCUSSION TITLE
Is cmp supposed to be used with sort/sorted?

### DISCUSSION THREAD

**Discussion by stutonk** [janet-lang/janet#1467] [Q&A]
`cmp` and functions like it that return either `-1`, `0`, or `1` when comparing two items -- are they meant to be used with `sort`/`sorted`? If not, is there a standard lib function or functions with which they're intended to be used?

**Comment by sogaiu** [comment]
> cmp and functions like it that return either -1, 0, or 1 when comparing two items -- are they meant to be used with sort/sorted

I don't think so.

IIUC, `sort` and `sorted` are meant to be used with things that return truthy / falsy values and all of `-1`, `0`, and `1` are truthy.

Here are some examples for `sort`:

```janet
(do
  (def arr @[2 0 1])
  (sort arr >)
  arr)
# =>
@[2 1 0]

(do
  (def arr @[[:fun 0] [:swim 2] [:play -1]])
  (sort arr (fn [[_ x] [_ y]]
              (< x y)))
  arr)
# =>
'@[(:play -1) (:fun 0) (:swim 2)]
```

---

`cmp` seems to be used as part of the internal macro [`do-compare`](https://github.com/janet-lang/janet/blob/a5d6b2283834422a9fa9e79b5c7ad9b932b52568/src/boot/boot.janet#L767-L773) (which is in turn used by [`compare`](https://github.com/janet-lang/janet/blob/a5d6b2283834422a9fa9e79b5c7ad9b932b52568/src/boot/boot.janet#L775-L782) and [`compare-reduce`](https://github.com/janet-lang/janet/blob/a5d6b2283834422a9fa9e79b5c7ad9b932b52568/src/boot/boot.janet#L784-L793) -- the latter which is then used by [`compare=`, ...,  `compare>=`](https://github.com/janet-lang/janet/blob/a5d6b2283834422a9fa9e79b5c7ad9b932b52568/src/boot/boot.janet#L795-L818) which are in turn used by the predicates [`zero?`, ..., `odd?`](https://github.com/janet-lang/janet/blob/a5d6b2283834422a9fa9e79b5c7ad9b932b52568/src/boot/boot.janet#L820-L825))...all of that stuff is listed under ["Polymorphic comparisons"](https://github.com/janet-lang/janet/blob/a5d6b2283834422a9fa9e79b5c7ad9b932b52568/src/boot/boot.janet#L765)...phew! (^^;

Possibly of interest might be [this document](https://janet-lang.org/docs/comparison.html).
